File: submissions/0114-flatten-binary-tree-to-linked-list/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void flatten(TreeNode root) {
        if (root == null) {
            return;
        } else {
            flatten(root.left);
            flatten(root.right);
            root.right = merge(root.left, root.right);
            root.left = null;
        }
    }
    
    private TreeNode merge(TreeNode flattenLeft, TreeNode flattenRight) {
        if (flattenLeft == null) {
            return flattenRight;    
        } else {
            if (flattenRight != null) {
                TreeNode current = flattenLeft;
                while(current.right != null) {
                    current = current.right;
                }
                current.right = flattenRight;
            }
            return flattenLeft;
        }
    }
}

File: submissions/0114-flatten-binary-tree-to-linked-list/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return None
        
        self.flattenAndReturnLastNode(root)

    def flattenAndReturnLastNode(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root: 
            return None
        if not root.left and not root.right:
            return root

        leftLast = self.flattenAndReturnLastNode(root.left)
        rightLast = self.flattenAndReturnLastNode(root.right)

        if not leftLast:
            return rightLast
        
        if not rightLast:
            root.right = root.left
            root.left = None
            return leftLast

        if leftLast and rightLast:
            leftLast.right = root.right
            root.right = root.left
            root.left = None
            return rightLast
        
